### ISSUE TITLE
TextCategorizer does not learn with integer labels

### ISSUE THREAD

**Issue by ValentinCalomme** [explosion/spaCy#4603]
The SpaCy text categorizer pipe does not learn if the training data has integer labels. For instance:

```py
annotations = {"cats": {0: True, 1: False}}
```

Passing this as annotations to the update method doesn't lead to any learning. The loss returned is 0 and the model's weights do not seem to be updated. On the other hand, the following annotations will work:

```py
annotations = {"cats": {"0": True, "1": False}}
```

To me, there are two possibilities:

1. Using string labels should be strictly enforced and an error should be thrown if this isn't the case

or

2. Integer labels should work as well as string labels due to duck typing

## Your Environment

* **spaCy version:** 2.2.2
* **Platform:** Windows-10-10.0.18362-SP0
* **Python version:** 3.7.5
* **Models:** nl

**Comment by honnibal** [comment]
Thanks for the report.

The correct solution is definitely 1: integers all correspond to a string in spaCy, so it'll lead to a lot of confusions if we accept arbitrary other integers. For instance, json encoding will turn an integer into a string, and spaCy will hash the string to some 64 bit integer that's not 0. So accepting integer labels will definitely be bad news.

**Comment by honnibal** [comment]
I've labelled this bug even though it's really more failure-to-be-friendly. I guess we don't have a category for that?

**Comment by ValentinCalomme** [comment]
Indeed, wasn't sure how to label it either. Should the desired behaviour be to raise an error or to simply cast all labels as strings if possible?

**Comment by adrianeboyd** [comment]
(We do have a category!)

**Comment by adrianeboyd** [comment]
Now with #4690, you shouldn't be able to add non-string labels at all, so you shouldn't run into this problem.
